consensus/propagation/reactor.go

```diff
@@ -135,3 +158,255 @@ func (blockProp *Reactor) Receive(chID byte, peer p2p.Peer, msgBytes []byte) {
+// node will request those parts.
+// The peer must always send the proposal before sending parts. If they did
+// not, this node must disconnect from them.
+// This method will:
+// - get the provided peer from the peer state
+// - get the proposal referenced in the haves message
+// - set the provided haves as the peer's haves
+// - if the returned parts from the proposal are complete, we return
+// - otherwise, we check if the sender has parts that we don't have.
+// - if they do, we check if we already requested those parts enough times (a limit will be defined)
+// of we already requested the parts from them.
+// - if so, we just gossip the haves to our connected peers.
+// - otherwise, we send the wants for the missing parts to that peer before broadcasting the haves.
+// - finally, we keep track of the want requests in the proposal state.
+func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts, bypassRequestLimit bool) {
+}
+
+// broadcastHaves gossips the provided have msg to all peers except to the
+// original sender. This should only be called upon receiving a new have for the
+// first time.
+//
+//nolint:unused
+func (blockProp *Reactor) broadcastHaves(haves *proptypes.HaveParts, from p2p.ID) {
+}
+
+// handleWants is called when a peer sends a want message. This is used to send
+// peers data that this node already has and store the wants to send them data
+// in the future.
+// This method will:
+// - get the provided peer from the state
+// - get the proposal from the proposal cache
+// - if the round provided in the wants is < 0, send the peer the partset header
+// - if we have the wanted parts, send them to that peer.
+// - if they want other parts that we don't have, store that in the peer state.
+func (blockProp *Reactor) handleWants(peer p2p.ID, wants *proptypes.WantParts) {
+}
+
+// sendPsh sends the partset header to the provided peer.
+// This method will:
+// - get the proposal with the provided height and round
+// - get the patset header from the block meta?
+// - send the partset header to the provided peer
+// TODO rename this Psh to something less Psh
+//
+//nolint:unused
+func (blockProp *Reactor) sendPsh(peer p2p.ID, height int64, round int32) bool {
+	return true
+}
+
+// HandleValidBlock is called when the node finds a peer with a valid block. If this
+// node doesn't have a block, it asks the sender for the portions that it
+// doesn't have.
+// This method will:
+// - get the provided peer from the peer state
+// - get the proposal referenced by the provided height and round
+// - if it has the proposal:
+//   - if the proposal is complete return
+//   - otherwise, create a new bit array with the size of the partset header total and fill it with true indices
+//   - broadcast the haves of that block
+//
+// - send the wants of all the block parts to the peer that sent it to us
+// - set the requests
+// - request all the previous blocks if any are missing
```

Comment:
in the future we might want to throttle this or something similar, but for now we can just do this